File: api/app/core/pdf.py
"""Statutory document PDFs via ReportLab (pure-Python, BSD, no system deps). Builders are pure:
they take a fully-computed data dict (exact paise, computed upstream by the payroll engine) and
return PDF bytes. They never compute money — they only render it."""

from __future__ import annotations

from io import BytesIO
from typing import Any

from reportlab.lib import colors
from reportlab.lib.pagesizes import A4
from reportlab.lib.styles import getSampleStyleSheet
from reportlab.platypus import Paragraph, SimpleDocTemplate, Spacer, Table, TableStyle

from app.core.legal import DISCLAIMER_TEXT
from app.core.money import Paise

_GRID = TableStyle(
    [
        ("GRID", (0, 0), (-1, -1), 0.5, colors.HexColor("#c9ced6")),
        ("BACKGROUND", (0, 0), (-1, 0), colors.HexColor("#1f2937")),
        ("TEXTCOLOR", (0, 0), (-1, 0), colors.white),
        ("FONTNAME", (0, 0), (-1, 0), "Helvetica-Bold"),
        ("FONTSIZE", (0, 0), (-1, -1), 9),
        ("ALIGN", (1, 0), (-1, -1), "RIGHT"),
        ("VALIGN", (0, 0), (-1, -1), "MIDDLE"),
    ]
)


def _inr(paise: Any) -> str:
    return Paise(int(paise)).format_inr()


def _disclaimer_footer(canvas: Any, _doc_template: Any) -> None:
    """WS10.4 — the byte-exact ``DISCLAIMER_TEXT`` on the foot of EVERY page of EVERY PDF this
    module emits (DISCLAIMERS.md placement policy). 8pt on a 9pt body — legible, not fine
    print."""
    canvas.saveState()
    canvas.setFont("Helvetica-Oblique", 8)
    canvas.drawCentredString(A4[0] / 2, 20, DISCLAIMER_TEXT)
    canvas.restoreState()


def _build(buf: BytesIO, title: str, elems: list[Any]) -> None:
    """One build path for every PDF, so the disclaimer footer cannot be forgotten on a new
    builder."""
    doc = SimpleDocTemplate(buf, pagesize=A4, title=title, author="Maisha-Mahsa")
    doc.build(elems, onFirstPage=_disclaimer_footer, onLaterPages=_disclaimer_footer)


def payslip_pdf(data: dict) -> bytes:
    """Monthly payslip. ``data``: company, employee_name, employee_code, period,
    earnings/deductions (list of [label, paise]), gross, total_deductions, net."""
    buf = BytesIO()
    styles = getSampleStyleSheet()
    elems: list[Any] = [
        Paragraph(f"{data['company']} — Payslip", styles["Title"]),
        Paragraph(
            f"{data['employee_name']} ({data['employee_code']}) &nbsp; · &nbsp; {data['period']}",
            styles["Normal"],
        ),
        Spacer(1, 10),
    ]
    earn = data["earnings"]
    ded = data["deductions"]
    rows = [["Earnings", "Amount (₹)", "Deductions", "Amount (₹)"]]
    for i in range(max(len(earn), len(ded))):
        e = earn[i] if i < len(earn) else ("", "")
        d = ded[i] if i < len(ded) else ("", "")
        rows.append([
            e[0], _inr(e[1]) if e[0] else "",
            d[0], _inr(d[1]) if d[0] else "",
        ])
    rows.append(["Gross", _inr(data["gross"]), "Total deductions", _inr(data["total_deductions"])])
    table = Table(rows, colWidths=[140, 110, 140, 110])
    table.setStyle(_GRID)
    net = Paragraph(f"<b>Net pay: {_inr(data['net'])}</b>", styles["Heading2"])
    elems += [table, Spacer(1, 12), net]
    _build(buf, f"Payslip {data['period']}", elems)
    return buf.getvalue()


def audit_pack_pdf(pack: dict) -> bytes:
    """Audit Pack artifact (§WS8.1). ``pack`` is the output of
    :func:`app.core.audit_pack.build_audit_pack` — every figure renders with its badge TEXT
    (VERIFIED/PENDING/BLOCKED, straight from the §0.4-gated badge, never re-decided here) and
    the pack integrity hash is embedded on the cover."""
    # no cycle: audit_pack does not import pdf
    from app.core.audit_pack import SECTION_ORDER, badge_text, section_title

    buf = BytesIO()
    styles = getSampleStyleSheet()
    integrity = pack["integrity"]
    elems: list[Any] = [
        Paragraph("Maisha-Mahsa — Audit Pack", styles["Title"]),
        Paragraph(
            f"Organisation: {pack['org_id']} &nbsp; · &nbsp; "
            f"Rules version: {pack['rules_version']}",
            styles["Normal"],
        ),
        Spacer(1, 6),
        Table(
            [["Integrity hash (SHA-256)", integrity["hash"]]],
            colWidths=[140, 360],
            style=_GRID,
        ),
        Spacer(1, 6),
        Paragraph(
            "Badge legend: VERIFIED = Mahsa independently recomputed this figure to the paisa; "
            "PENDING = not yet independently recomputed; BLOCKED = recompute mismatch. "
            "No figure is marked VERIFIED without a Mahsa recomputation (Prime Directive §0.4).",
            styles["Italic"],
        ),
        Spacer(1, 12),
    ]
    for name in SECTION_ORDER:
        elems.append(Paragraph(section_title(name), styles["Heading2"]))
        figures = pack["sections"][name]
        if figures:
            rows = [["Particulars", "Amount (₹)", "Badge", "Evidence"]]
            for f in figures:
                rows.append([
                    Paragraph(f["label"], styles["BodyText"]),
                    _inr(f["value_paise"]),
                    badge_text(f["badge"]),
                    Paragraph(f["evidence_ref"], styles["BodyText"]),
                ])
                # CITE.P1-1 (SPEC-MEMCITE-1.0 §B4.2): the figure's cell-level source anchors —
                # excerpt + resolution state verbatim from the sealed pack. A MOVED or BROKEN
                # citation prints as such; the export never claims RESOLVED for it.
                for a in f.get("anchors", []):
                    rows.append([
                        Paragraph(f"source: {a.get('excerpt', '')}", styles["BodyText"]),
                        "",
                        str(a.get("resolution", "")).upper(),
                        Paragraph(a.get("note") or "", styles["BodyText"]),
                    ])
            table = Table(rows, colWidths=[210, 100, 70, 120])
            table.setStyle(_GRID)
            elems.append(table)
        note = pack["section_notes"].get(name)
        if note:
            elems.append(Spacer(1, 4))
            elems.append(Paragraph(note, styles["Italic"]))
        elems.append(Spacer(1, 10))
    if pack["pending_sections"]:
        elems.append(
            Paragraph(
                "Not yet included: " + ", ".join(pack["pending_sections"]) + ".",
                styles["Italic"],
            )
        )
    _build(buf, "Audit Pack", elems)
    return buf.getvalue()


def form16_pdf(data: dict) -> bytes:
    """Form 16 — Part B (salary TDS certificate). ``data``: company, tan, employee_name, pan,
    financial_year, assessment_year, rows (list of [label, paise]), total_tax_deducted."""
    buf = BytesIO()
    styles = getSampleStyleSheet()
    elems: list[Any] = [
        Paragraph("FORM NO. 16 — Part B", styles["Title"]),
        Paragraph(
            "[See rule 31(1)(a)] · Certificate under section 203 of the Income-tax Act, 1961",
            styles["Normal"],
        ),
        Spacer(1, 8),
        Paragraph(
            f"Employer: {data['company']} (TAN: {data.get('tan') or '—'})<br/>"
            f"Employee: {data['employee_name']} (PAN: {data.get('pan') or '—'})<br/>"
            f"Financial Year: {data['financial_year']} &nbsp; · &nbsp; "
            f"Assessment Year: {data['assessment_year']}",
            styles["Normal"],
        ),
        Spacer(1, 10),
    ]
    rows = [["Particulars", "Amount (₹)"]]
    rows += [[label, _inr(amount)] for label, amount in data["rows"]]
    rows.append(["Total tax deducted (TDS)", _inr(data["total_tax_deducted"])])
    table = Table(rows, colWidths=[360, 140])
    table.setStyle(_GRID)
    elems += [
        table,
        Spacer(1, 12),
        Paragraph(
            "This is Part B (salary breakup &amp; tax computation) generated from payroll. "
            "Part A (TDS deposited) is issued via TRACES.",
            styles["Italic"],
        ),
    ]
    _build(buf, f"Form 16 {data['financial_year']}", elems)
    return buf.getvalue()

File: api-nest/src/pdf/pdf.ts
/**
 * Statutory document PDFs. Port of api/app/core/pdf.py (which used ReportLab). Builders are pure:
 * they take a fully-computed data dict (exact paise, computed upstream by the payroll engine) and
 * return PDF bytes. They never compute money — they only render it (via formatInr).
 */
import { join } from 'path';
import PDFDocument from 'pdfkit';

import { formatInr } from '../common/money';

// DejaVu Sans is bundled (assets/fonts/) because it carries the ₹ glyph (U+20B9) that pdfkit's
// built-in Helvetica lacks. Resolved relative to the compiled file so it works in the container.
const FONT_DIR = join(__dirname, '..', '..', 'assets', 'fonts');
const FONT_REGULAR = join(FONT_DIR, 'DejaVuSans.ttf');
const FONT_BOLD = join(FONT_DIR, 'DejaVuSans-Bold.ttf');

/** A new PDF document with the ₹-capable fonts registered as `body` / `bold`. */
function newDoc(title: string): PDFKit.PDFDocument {
  const doc = new PDFDocument({ size: 'A4', margin: 48, info: { Title: title, Author: 'Maisha-Mahsa' } });
  doc.registerFont('body', FONT_REGULAR);
  doc.registerFont('bold', FONT_BOLD);
  doc.font('body');
  return doc;
}

function toBuffer(doc: PDFKit.PDFDocument): Promise<Buffer> {
  return new Promise((resolve, reject) => {
    const chunks: Buffer[] = [];
    doc.on('data', (c: Buffer) => chunks.push(c));
    doc.on('end', () => resolve(Buffer.concat(chunks)));
    doc.on('error', reject);
    doc.end();
  });
}

const inr = (paise: unknown) => formatInr(Math.trunc(Number(paise)));

export interface PayslipData {
  company: string;
  employee_name: string;
  employee_code: string;
  period: string;
  earnings: [string, number][];
  deductions: [string, number][];
  gross: number;
  total_deductions: number;
  net: number;
}

/** Monthly payslip PDF. */
export function payslipPdf(data: PayslipData): Promise<Buffer> {
  const doc = newDoc(`Payslip ${data.period}`);
  doc.font('bold').fontSize(18).text(`${data.company} — Payslip`);
  doc.font('body').moveDown(0.3).fontSize(10).fillColor('#444').text(`${data.employee_name} (${data.employee_code})  ·  ${data.period}`);
  doc.moveDown(1).fillColor('#000');

  const rows = Math.max(data.earnings.length, data.deductions.length);
  doc.font('bold').fontSize(11).text('Earnings', 48, doc.y, { continued: false });
  const startY = doc.y;
  // Two side-by-side columns: earnings (left), deductions (right).
  doc.font('body').fontSize(10);
  for (let i = 0; i < rows; i++) {
    const y = startY + 6 + i * 16;
    const e = data.earnings[i];
    const d = data.deductions[i];
    if (e) doc.text(`${e[0]}`, 48, y).text(inr(e[1]), 200, y, { width: 90, align: 'right' });
    if (d) doc.text(`${d[0]}`, 320, y).text(inr(d[1]), 470, y, { width: 90, align: 'right' });
  }
  const totalsY = startY + 6 + rows * 16 + 8;
  doc.font('bold').fontSize(10.5).text('Gross', 48, totalsY).text(inr(data.gross), 200, totalsY, { width: 90, align: 'right' });
  doc.text('Total deductions', 320, totalsY).text(inr(data.total_deductions), 470, totalsY, { width: 90, align: 'right' });
  doc.font('bold').moveDown(3).fontSize(14).text(`Net pay: ${inr(data.net)}`, 48);
  return toBuffer(doc);
}

export interface Form16Data {
  company: string;
  tan?: string | null;
  employee_name: string;
  pan?: string | null;
  financial_year: string;
  assessment_year: string;
  rows: [string, number][];
  total_tax_deducted: number;
}

/** Form 16 — Part B (salary TDS certificate). */
export function form16Pdf(data: Form16Data): Promise<Buffer> {
  const doc = newDoc(`Form 16 ${data.financial_year}`);
  doc.font('bold').fontSize(16).text('FORM NO. 16 — Part B');
  doc.font('body').moveDown(0.3).fontSize(9).fillColor('#444').text('[See rule 31(1)(a)] · Certificate under section 203 of the Income-tax Act, 1961');
  doc.moveDown(0.8).fillColor('#000').fontSize(10);
  doc.text(`Employer: ${data.company} (TAN: ${data.tan || '—'})`);
  doc.text(`Employee: ${data.employee_name} (PAN: ${data.pan || '—'})`);
  doc.text(`Financial Year: ${data.financial_year}  ·  Assessment Year: ${data.assessment_year}`);
  doc.moveDown(1);

  const startY = doc.y;
  doc.fontSize(10);
  const allRows: [string, number][] = [...data.rows, ['Total tax deducted (TDS)', data.total_tax_deducted]];
  allRows.forEach(([label, amount], i) => {
    const y = startY + i * 18;
    const last = i === allRows.length - 1;
    doc.font(last ? 'bold' : 'body').text(label, 48, y).text(inr(amount), 400, y, { width: 110, align: 'right' });
  });
  doc.font('body').fontSize(9).fillColor('#444').text(
    'This is Part B (salary breakup & tax computation) generated from payroll. Part A (TDS deposited) is issued via TRACES.',
    48,
    startY + allRows.length * 18 + 20,
  );
  return toBuffer(doc);
}
